design/validation-spi.md

```diff
@@ -0,0 +1,293 @@
+with a custom implementation that provides a narrow API. Additionally custom extensions should be easier and provide 
+a better integration with the existing Keycloak infrastructure. Further more dynamic validation logic lookups 
+and configuration should be possible and easy to do. 
+
+Note that it is still possible to provide a validation/validator implementation that leverages bean validation. 
+Also, bean-validation still works quite well for declarative validation in JAX-RS endpoints.
+
+### Fully fledged Validation SPI
+
+An earlier attempt at a unified validation SPI can be found here [KEYCLOAK-2045 Add support for flexible Validation](https://github.com/keycloak/keycloak/pull/7324) and was discussed on the [keycloak-dev mailing list](https://groups.google.com/g/keycloak-dev/c/-XjQu7rn56s).
+
+After some discussion, it became apparent that this proposal provides a very comprehensive validation framework that 
+offers numerous functionalities that go beyond the requirements of user profile validation and other validations in Keycloak.
+
+However, the API was very complex and not easy to understand. Therefore, it was agreed that @thomasdarimont would revise 
+the proposal again and present it in a smaller form. This has resulted in the new "Simple Validation SPI" proposal.
+
+## Proposed Validation SPI
+
+### PoC for new Validator SPI
+
+The Proof of concept implementation can be found here in the PR [KEYCLOAK-2045 Simple Validation API #7887](https://github.com/keycloak/keycloak/pull/7887).
+
+Some usage examples can be found here in the [ValidatorTest](https://github.com/thomasdarimont/keycloak/blob/issue/KEYCLOAK-2045-Simple-Validation-SPI/server-spi-private/src/test/java/org/keycloak/validate/ValidatorTest.java).
+
+### Core API
+The proposal focus around the following types:
+1. `Validator`: `Provider` interface for validation mechanics.
+1. `ValidationContext`: Holds state of a sequence of validations.
+1. `ValidationResult`: Denotes the outcome of a validation.
+1. `ValidatorConfig`: Denotes the configuration for a Validator. A typed wrapper around a `Map<String,Object>`.
+1. `ValidationError`: Represents an error found during validation.
+1. `ValidatorLookup`: Helper class to provide lookups for built-in and user provided validations.
+
+### Support API
+The following types provide the integration with the Keycloak infrastructure:
+1. `ValidatorFactory`: `ProviderFactory` interface for custom validation contributions.
+1. `ValidatorSPI`: Provides the `validator` SPI.
+1. `CompactValidator`: Convenience class for built-in validations.
+1. `BuiltinValidators`: Denotes a registry for internal built-in validations.
+
+The proposed package name is `org.keycloak.validation`. Note the current PR [KEYCLOAK-2045 Simple Validation API](https://github.com/keycloak/keycloak/pull/7887) uses the package name `org.keycloak.validate` to provide a transition path for the current API.
```

Comment:
This list needs big elaboration. How do these classes work together? How does it compare to to JSR 380?

Reply:
The proposed Validation SPI is simpler than JSR 380 and deliberately provides less features. It it works without reflection and does not use annotations. The extension model is aligned with Keycloak's extension mechanism.